File: Sources/SpotHeroEmailValidator/SHEmailValidationTextField.m
//  Copyright © 2019 SpotHero, Inc. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#include <TargetConditionals.h>

#if TARGET_OS_UIKITFORMAC || !TARGET_OS_OSX

#if ! __has_feature(objc_arc)
#error This file must be compiled with ARC. Either turn on ARC for the project or use -fobjc-arc flag
#endif

#import "include/SHEmailValidationTextField.h"
#import "include/SHAutocorrectSuggestionView.h"
#import <SpotHeroEmailValidator/SpotHeroEmailValidator-Swift.h>

@interface SHEmailValidationTextField () <AutocorrectSuggestionViewDelegate>

@property (nonatomic, strong) SHAutocorrectSuggestionView *suggestionView;
@property (nonatomic, strong) SpotHeroEmailValidator *emailValidator;
@property (nonatomic, strong) EmailTextFieldDelegate *delegateProxy;
@property (nonatomic, strong) NSMutableDictionary *messageDictionary;

@end

@implementation SHEmailValidationTextField

- (id)init
{
    self = [super init];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (void)commonInit
{
    self.delegateProxy = [[EmailTextFieldDelegate alloc] initWithTarget:self];
    self.delegate = self.delegateProxy;
    self.emailValidator = [SpotHeroEmailValidator shared];
    self.autocapitalizationType = UITextAutocapitalizationTypeNone;
    self.keyboardType = UIKeyboardTypeEmailAddress;
    self.autocorrectionType = UITextAutocorrectionTypeNo;
    self.messageDictionary = [NSMutableDictionary dictionary];
    self.defaultErrorMessage = @"Please enter a valid email address";
    self.messageForSuggestion = @"Did you mean";

    self.bubbleFillColor = [SHAutocorrectSuggestionView defaultFillColor];
    self.bubbleTitleColor = [SHAutocorrectSuggestionView defaultTitleColor];
    self.bubbleSuggestionColor = [SHAutocorrectSuggestionView defaultSuggestionColor];
}

- (void)setDelegate:(id<UITextFieldDelegate>)delegate
{
    if ([delegate isKindOfClass:[EmailTextFieldDelegate class]]) {
        [super setDelegate:delegate];
    } else {
        self.delegateProxy.subDelegate = delegate;
    }
}

- (void)validateInput
{
    if (self.text.length > 0) {
        NSError *error;
        
        SHValidationResult *validationResult = [self.emailValidator validateAndAutocorrectWithEmailAddress:self.text error:&error];

        if (error) {
            self.validationError = error;

            NSString *message = self.messageDictionary[@(error.code)];
            if (!message) {
                message = self.defaultErrorMessage;
            }
            self.suggestionView = [SHAutocorrectSuggestionView showFromView:self title:message autocorrectSuggestion:nil withSetupBlock:^(SHAutocorrectSuggestionView *view) {
                view.fillColor = self.bubbleFillColor;
                view.titleColor = self.bubbleTitleColor;
                view.suggestionColor = self.bubbleSuggestionColor;
            }];
            self.suggestionView.delegate = self;
        } else {
            self.validationError = nil;

            if (validationResult.autocorrectSuggestion) {
                self.suggestionView = [SHAutocorrectSuggestionView showFromView:self title:self.messageForSuggestion autocorrectSuggestion:validationResult.autocorrectSuggestion withSetupBlock:^(SHAutocorrectSuggestionView *view) {
                    view.fillColor = self.bubbleFillColor;
                    view.titleColor = self.bubbleTitleColor;
                    view.suggestionColor = self.bubbleSuggestionColor;
                }];
                self.suggestionView.delegate = self;
            }
        }
    }
}

#if !TARGET_OS_TV
- (void)hostWillAnimateRotationToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    [self.suggestionView updatePosition];
}
#endif

- (void)dismissSuggestionView
{
    [self.suggestionView dismiss];
}

- (void)setMessage:(NSString *)message forErrorCode:(NSInteger)errorCode
{
    self.messageDictionary[@(errorCode)] = message;
}

- (void)setBubbleFillColor:(UIColor *)bubbleFillColor
{
    _bubbleFillColor = bubbleFillColor;
    self.suggestionView.fillColor = bubbleFillColor;
}

- (void)setBubbleTitleColor:(UIColor *)bubbleTitleColor
{
    _bubbleTitleColor = bubbleTitleColor;
    self.suggestionView.titleColor = bubbleTitleColor;
}

- (void)setBubbleSuggestionColor:(UIColor *)bubbleSuggestionColor
{
    _bubbleSuggestionColor = bubbleSuggestionColor;
    self.suggestionView.suggestionColor = bubbleSuggestionColor;
}

#pragma mark
#pragma mark - AutocorrectSuggestionViewDelegate methods
- (void)suggestionView:(SHAutocorrectSuggestionView *)suggestionView wasDismissedWithAccepted:(BOOL)accepted
{
    if (accepted) {
        self.text = suggestionView.suggestedText;
    }
}

@end

#endif

File: Sources/SpotHeroEmailValidator/SHEmailValidationTextField.swift
// Copyright © 2022 SpotHero, Inc. All rights reserved.

#if canImport(UIKit)

    import Foundation
    import UIKit

    public class SHEmailValidationTextField: UITextField {
        var suggestionView: SHAutocorrectSuggestionView?
        var emailValidator: SpotHeroEmailValidator?
        var delegateProxy: EmailTextFieldDelegate?
        var messageDictionary: [Int: String]?
    
        public var defaultErrorMessage: String?
        public var messageForSuggestion: String?
        public var validationError: Error?
    
        public var bubbleFillColor: UIColor? {
            didSet {
                self.suggestionView?.fillColor = self.bubbleFillColor
            }
        }
    
        public var bubbleTitleColor: UIColor? {
            didSet {
                self.suggestionView?.titleColor = self.bubbleTitleColor
            }
        }
    
        public var bubbleSuggestionColor: UIColor? {
            didSet {
                self.suggestionView?.suggestionColor = self.bubbleSuggestionColor
            }
        }
    
        override public var delegate: UITextFieldDelegate? {
            get {
                return super.delegate
            }
            set {
                if newValue is EmailTextFieldDelegate {
                    super.delegate = newValue
                } else {
                    self.delegateProxy?.subDelegate = newValue
                }
            }
        }
    
        public convenience init() {
            self.init(frame: .zero)
        }
    
        public required init?(coder: NSCoder) {
            super.init(coder: coder)
        
            self.initialize()
        }
    
        override public init(frame: CGRect) {
            super.init(frame: frame)
        
            self.initialize()
        }
    
        private func initialize() {
            self.delegateProxy = EmailTextFieldDelegate(target: self)
            self.delegate = self.delegateProxy
            self.emailValidator = SpotHeroEmailValidator.shared
            self.autocapitalizationType = .none
            self.keyboardType = .emailAddress
            self.autocorrectionType = .no
            self.messageDictionary = [:]
            self.defaultErrorMessage = "Please enter a valid email address"
            self.messageForSuggestion = "Did you mean"
        
            self.bubbleFillColor = SHAutocorrectSuggestionView.defaultFillColor()
            self.bubbleTitleColor = SHAutocorrectSuggestionView.defaultTitleColor()
            self.bubbleSuggestionColor = SHAutocorrectSuggestionView.defaultSuggestionColor()
        }
    
        public func dismissSuggestionView() {
            self.suggestionView?.dismiss()
        }
    
        public func validateInput() {
            guard let text = self.text, !text.isEmpty else {
                return
            }
        
            // TODO: This method can be drastically reduced, lots of redundant code
            do {
                let validationResult = try self.emailValidator?.validateAndAutocorrect(emailAddress: text)
            
                self.validationError = nil
            
                if let autocorrectSuggestion = validationResult?.autocorrectSuggestion {
                    self.suggestionView = SHAutocorrectSuggestionView.show(from: self,
                                                                           title: self.messageForSuggestion,
                                                                           autocorrectSuggestion: autocorrectSuggestion,
                                                                           withSetupBlock: { [weak self] view in
                                                                               view?.fillColor = self?.bubbleFillColor
                                                                               view?.titleColor = self?.bubbleTitleColor
                                                                               view?.suggestionColor = self?.bubbleSuggestionColor
                                                                           })
                
                    self.suggestionView?.delegate = self
                }
            } catch {
                self.validationError = error
            
                var message = self.messageDictionary?[(error as NSError).code]
            
                if message?.isEmpty == true {
                    message = self.defaultErrorMessage
                }

                self.suggestionView = SHAutocorrectSuggestionView.show(from: self,
                                                                       title: message,
                                                                       autocorrectSuggestion: nil,
                                                                       withSetupBlock: { [weak self] view in
                                                                           view?.fillColor = self?.bubbleFillColor
                                                                           view?.titleColor = self?.bubbleTitleColor
                                                                           view?.suggestionColor = self?.bubbleSuggestionColor
                                                                       })
            
                self.suggestionView?.delegate = self
            }
        }

        public func setMessage(_ message: String, forErrorCode errorCode: Int) {
            self.messageDictionary?[errorCode] = message
        }
    }

    extension SHEmailValidationTextField: AutocorrectSuggestionViewDelegate {
        public func suggestionView(_ suggestionView: SHAutocorrectSuggestionView, wasDismissedWithAccepted accepted: Bool) {
            if accepted {
                self.text = suggestionView.suggestedText
            }
        }
    }

    #if !os(tvOS)

        public extension SHEmailValidationTextField {
            /// Must be called manually when animating a rotation so that the suggestion view is reoriented properly.
            func hostWillAnimateRotation(to interfaceOrientation: UIInterfaceOrientation) {
                self.suggestionView?.updatePosition()
            }
        }

    #endif

#endif
